src/plugins/dist_matrixops/all_gather.cpp

```diff
@@ -0,0 +1,383 @@
+            case 0:
+                return detail::concatenate2d_axis0<T>(
+                    std::move(args), name, codename);
+            case 1:
+                return detail::concatenate2d_axis1<T>(
+                    std::move(args), name, codename);
+            default:
+                break;
+            }
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "dist_matrixops::primitives::"
+                "all_gather::detail::concatenate2d",
+                util::generate_error_message(
+                    "axis is out of bounds of dimension", name, codename));
+        }
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    execution_tree::primitive_argument_type all_gather::all_gather2d(
+        ir::node_data<T>&& arr,
+        execution_tree::localities_information&& locs) const
+    {
+        using namespace execution_tree;
+        auto m = arr.matrix();
+        blaze::DynamicMatrix<T> res_value(m.rows(), m.columns());
+        res_value = m;
+        // use hpx::all_gather to get a vector of values
+        auto p = hpx::all_gather(
+            ("all_gather_" + locs.annotation_.name_).c_str(),
+            res_value, locs.locality_.num_localities_,
+            std::size_t(-1),
+            locs.locality_.locality_id_)
+                .get();
+
+        // row and column dimensions of the whole array
+        std::size_t rows_dim, cols_dim;
+        rows_dim = locs.rows();
+        cols_dim = locs.columns();
+
+        // check the tiling type is column-tiling or row-tiling
+        std::int64_t axis; // along with the array will be joined
+        if (m.rows() == rows_dim)
+        {
+            // column-tiling
+            axis = 1;
+        }
+        else if (m.columns() == cols_dim)
+        {
+            // row-tiling
+            axis = 0;
+        }
+        else
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "all_gather::detail::all_gather2d",
+                generate_error_message(
+                    "invalid tiling_type. The tiling_type can"
+                    "be `row` or `column`"));
+        }
+
+        primitive_arguments_type ops;
+        ops.reserve(p.size());
+        for (auto && op : p)
+        {
+            ops.push_back(
+                primitive_argument_type{ir::node_data<T>{std::move(op)}});
+        }
+
+        return detail::concatenate2d<T>(std::move(ops), axis, name_, codename_);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    execution_tree::primitive_argument_type all_gather::all_gather2d(
+        execution_tree::primitive_argument_type&& arr) const
+    {
+        using namespace execution_tree;
+
+        execution_tree::localities_information locs =
+            extract_localities_information(arr, name_, codename_);
+
+        std::size_t ndim = locs.num_dimensions();
+        if (ndim > 2 || ndim < 1)
```

Comment:
It might not be necessary, but it would be good if it works with empty arrays too

Reply:
Yes. It makes sense to consider empty arrays. I assume what I need to add is that ndim can also be equal to 0